### PR TITLE
Calendar Page, Step and Feature completed

### PR DESCRIPTION
**Pull request by LanaDIO**
Please review

### CODE REVIEW

File: apps/tenant-management-webapp-e2e/src/integration/calendar/calendar.page.ts [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+  {
+    return cy.xpath('//goa-modal[@data-testid="add-calendar-modal" and @open="true"]//*[@slot="heading"]');
+  }
+
+  calendarModalCancelButton()
+  {
+    return cy.get('[data-testid="calendar-modal-cancel"]');
+  }
+
+  calendarModalSaveButton()
```
Comment (howard-zhang) [inline apps/tenant-management-webapp-e2e/src/integration/calendar/calendar.page.ts:20]: Can this page object be shared with editCalendarModalSaveButton()? It's named calendarModalSaveButton(), and user would assume it's for any save button in any calendar modal.

File: apps/tenant-management-webapp-e2e/src/integration/calendar/calendar.page.ts [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+  calendarEditButton(rowNumber) {
+    return cy.xpath(`(//table[@data-testid="calendar-table"]//*[contains(@data-testid, "calendar-edit")])[${rowNumber}]`);
+  }
+
+  calendarDeleteButton(rowNumber) {
+    return cy.xpath(`(//table[@data-testid="calendar-table"]//*[contains(@data-testid, "delete-icon")])[${rowNumber}]`);
+  }
+
+  editCalendarModal() {
```
Comment (howard-zhang) [inline apps/tenant-management-webapp-e2e/src/integration/calendar/calendar.page.ts:74]: Looks like your edit modal page objects have add-calendar-modal ids. Can they be shared with the corresponding add modal page objects?

File: apps/tenant-management-webapp-e2e/src/integration/calendar.feature [hunk 3]
```diff
@@ -16,4 +16,34 @@ Feature: Calendar
...
+  @TEST_CS-330 @REQ_CS-580 @REQ_CS-1652 @REQ_CS-1709 @regression
+  Scenario: As a tenant admin, I can add, edit and delete calendars in the tenant admin view
+    Given a tenant admin user is on tenant admin page
+    When the user selects the "Calendar" menu item
+    Then the user views the "Calendar service" overview content "The calendar service provides"
+    When the user clicks Add calendar button on overview tab
+    Then the user views Add calendar modal
+    When the user clicks Cancel button in Add calendar modal
+    Then the user views Calendar tab table header
+    When the user clicks Add calendar button
```
Comment (howard-zhang) [inline apps/tenant-management-webapp-e2e/src/integration/calendar.feature:30]: Please add "on calendars page" to be more specific.

File: apps/tenant-management-webapp-e2e/src/integration/calendar/calendar.step.ts [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+    .should('contain', errorMsg);
+});
+
+When('the user enters {string}, {string}, {string} in Add calendar modal',
+  function (name, desc, role) {
+    cy.viewport(1920, 1080);
+    calendarObj.addCalendarModalNameField().shadow().find('input').clear().type(name, { delay: 100, force: true });
+    calendarObj.addCalendarModalDescriptionField().shadow().find('.goa-textarea').clear().type(desc, { force: true });
+    // Select roles or client roles
+ const roles = role.split(',');
```
Comment (howard-zhang) [inline apps/tenant-management-webapp-e2e/src/integration/calendar/calendar.step.ts:63]: Indentation seems odd. Do you have Prettier - Code formatter vs code extension installed? VS code should automatically add indentation for you.

**Review by howard-zhang — CHANGES_REQUESTED** [review]

File: apps/tenant-management-webapp-e2e/src/integration/calendar/calendar.page.ts [hunk 5]
```diff
@@ -0,0 +1,89 @@
+class CalendarPage {
+
+  addCalendarOverviewTabBtn() {
+    return cy.get('[data-testid="overall-calendar-add-btn"]');
+  }
+  addCalendarBtn() {
```
[the remark this answers is not in the document]
Comment (LanaDIO) [reply apps/tenant-management-webapp-e2e/src/integration/calendar/calendar.page.ts:6]: added

File: apps/tenant-management-webapp-e2e/src/integration/calendar.feature (hunk 3 above)
Comment (LanaDIO) [reply apps/tenant-management-webapp-e2e/src/integration/calendar.feature:30]: added

File: apps/tenant-management-webapp-e2e/src/integration/calendar/calendar.page.ts (hunk 2 above)
Comment (LanaDIO) [reply apps/tenant-management-webapp-e2e/src/integration/calendar/calendar.page.ts:74]: Yes one can be used for both.

File: apps/tenant-management-webapp-e2e/src/integration/calendar.feature [hunk 6]
```diff
@@ -16,4 +16,34 @@ Feature: Calendar
...
+    When the user enters "auto-test-1A $" in name field in calendar modal
+    Then the user views the error message of "Allowed characters are: a-z, A-Z, 0-9, -, [space]" on namespace in calendar modal
+    # Validate data
+    When the user enters "autotest-addcalendar", "autotest calendar desc", "auto-test-role1, urn:ads:autotest:chat-service:chat-admin" in Add calendar modal
+    And the user clicks Save button in Add calendar modal
+    Then the user "views" the calendar of "autotest-addcalendar", "autotest calendar desc", "auto-test-role1, urn:ads:autotest:chat-service:chat-admin"
+    # Edit
+    When the user clicks "Edit" button for the calendar of "autotest-addcalendar", "autotest calendar desc", "auto-test-role1, urn:ads:autotest:chat-service:chat-admin"
+    Then the user views Edit calendar modal
+    When the user enters "autotest calendar desc edit" as description in Edit calendar modal
```
[the remark this answers is not in the document]
Comment (LanaDIO) [reply apps/tenant-management-webapp-e2e/src/integration/calendar.feature:42]: Added edit for the roles

File: apps/tenant-management-webapp-e2e/src/integration/calendar.feature [hunk 7]
```diff
@@ -16,4 +16,35 @@ Feature: Calendar
...
+    Then the user select "auto-test-role2" as role in Edit calendar modal
```
Comment (howard-zhang) [inline apps/tenant-management-webapp-e2e/src/integration/calendar.feature:43]: Could you please look at what we have for editing an item with roles in other services? The existing implementation for editing for roles is to clear all checkboxes and select the ones passed in as the parameter for roles. What you have is just selecting another unselected checkbox, assuming no touch on other checked ones and the ones you try to select are not selected yet.

Also I think we should have a single step to handle editing all possible fields including description and roles, like what we did in other services.

File: apps/tenant-management-webapp-e2e/src/integration/calendar/calendar.step.ts [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+Then('the user views Edit calendar modal', function () {
+  calendarObj.editCalendarModal().should('exist');
+});
+
+When('the user enters {string} as description in Edit calendar modal', function (description) {
+  calendarObj.editCalendarModalDescriptionField().shadow().find('.goa-textarea').clear().type(description, { force: true });
+});
+
+When('the user select {string} as role in Edit calendar modal', function (role) {
+  // calendarObj.editCalendarModalDescriptionField().shadow().find('.goa-textarea').clear().type(description, { force: true });
```
Comment (howard-zhang) [inline apps/tenant-management-webapp-e2e/src/integration/calendar/calendar.step.ts:202]: Please remove the not-needed code.

File: apps/tenant-management-webapp-e2e/src/integration/calendar/calendar.step.ts [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+  const roles = role.split(',');
+    for (let i = 0; i < roles.length; i++) {
+      if (roles[i].includes(':')) {
```
Comment (howard-zhang) [inline apps/tenant-management-webapp-e2e/src/integration/calendar/calendar.step.ts:205]: Missing the steps to check if the checkbox is already checked or not. The assumption of the current checkbox being in unchecked status adds a limitation of calling this step. What we usually do in other services is to clear all role checkboxes and select the ones (complete list) passed in as role parameter.

**Review by howard-zhang — CHANGES_REQUESTED** [review]

File: apps/tenant-management-webapp-e2e/src/integration/calendar/calendar.step.ts (hunk 8 above)
Comment (LanaDIO) [reply apps/tenant-management-webapp-e2e/src/integration/calendar/calendar.step.ts:202]: combined into one step

File: apps/tenant-management-webapp-e2e/src/integration/calendar.feature (hunk 7 above)
Comment (LanaDIO) [reply apps/tenant-management-webapp-e2e/src/integration/calendar.feature:43]: Added clear all checkboxes

**Review by howard-zhang — APPROVED** [review]

**Pull request merged**